feat: allow customizing job handling executor separately

chillleader opened the pull request:
## Description

<!-- Describe the goal and purpose of this PR. -->

**Problem:** Currently job workers can only be customized with one `ScheduledExecutorService` executor that is used for everything inside the worker - scheduling polling requests and handling jobs. This is a limitation, see linked issue below.

**Goal:** Allow customizing job workers with both a `ScheduledExecutorService` for scheduling and an `ExecutorService` for handling jobs - both in polling and streaming modes.

**Solution description:**
- Job worker code responsible for handling jobs is already written in a way that differentiates the two executors - the problem is that this config is not exposed to the users.
- I modified `ExecutorResource` so that it defines two executors - as a minimum, the calling code must still provide at least a `ScheduledExecutorService`, but can also provide an additional job executor.
- Exposed this via config in `CamundaClientBuilder`.
- Spring Boot Starter already allows customization by defining a `CamundaClientExecutorService` bean.
- Updated the metrics implementation, moving the meter registry binding into `CamundaClientExecutorService` to make sure any user-defined executor coming from the Spring context is metered by default.

## Checklist

<!--- Please delete options that are not relevant. Boxes should be checked by reviewer. -->
- [ ] Enable backports when necessary (fex. [for bug fixes](https://github.com/camunda/camunda/blob/main/CONTRIBUTING.md#backporting-changes) or [for CI changes](https://github.com/camunda/camunda/wiki/CI-&-Automation#when-to-backport-ci-changes)).

## Related issues

closes #39304

chillleader committed:
feat: allow customizing job handling executor separately

chillleader committed:
feat: add tests for custom job handling executor

megglos commented on clients/java/src/main/java/io/camunda/client/CamundaClientConfiguration.java:
```diff
@@ -141,6 +142,10 @@ public interface CamundaClientConfiguration {
    */
   boolean ownsJobWorkerExecutor();
 
+  ExecutorService jobHandlingExecutor();
```
❌  for consistency let's add the javadoc to link to the equivalent builder method

chillleader committed:
fix: add logging to ExecutorResource

chillleader committed:
refactor: move executor service metrics into CamundaClientExecutorService

This change makes sure user-defined job executors from the Spring context are always metered.
Additionally, it changes the impl to use the wrapped executor which adds more metrics
(e.g. task timing/counters)

chillleader committed:
fix: propagate spring-defined executor correctly

This change adds correct propagation of the user-defined executor service
from the spring context to the camunda client configuration.

megglos commented on clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/jobhandling/CamundaClientExecutorService.java:
```diff
@@ -27,29 +35,81 @@
     return scheduledExecutorService;
   }
 
-  public static CamundaClientExecutorService createDefault() {
```
❌  why remove that default no args variant? This caused further changes that don't seem necessary?

chillleader committed:
fix: remove nullable since not used anywhere else

This way we avoid introducing the new annotations /
adding new dependencies

megglos reviewed (changes requested):
Thanks for contributing this feature! A couple of comments I'd like to get addressed or align on, thus requesting changes.

megglos commented on clients/camunda-spring-boot-starter/src/main/java/io/camunda/client/spring/configuration/CamundaClientAllAutoConfiguration.java:
```diff
@@ -59,17 +60,18 @@ public CamundaClientAllAutoConfiguration(final CamundaClientProperties camundaCl
   @ConditionalOnMissingBean
-  public CamundaClientExecutorService camundaClientExecutorService() {
+  public CamundaClientExecutorService camundaClientExecutorService(
+      @Autowired(required = false) final MeterRegistry meterRegistry) {
```
❌  this will introduce a classpath dependency on micrometer for all setups, that's why the metric setup was in a dedicated config class which was conditional on the presence of the micrometer class `@ConditionalOnClass(MeterRegistry.class)`

megglos replied:
See the failing CPT tests:
```
Caused by: java.lang.IllegalStateException: Failed to introspect Class [io.camunda.client.spring.configuration.CamundaClientAllAutoConfiguration] from ClassLoader [jdk.internal.loader.ClassLoaders$AppClassLoader@76ed5528]
	at org.springframework.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:483) ~[spring-core-6.2.11.jar:6.2.11]
	at org.springframework.util.ReflectionUtils.findMethod(ReflectionUtils.java:238) ~[spring-core-6.2.11.jar:6.2.11]
	at org.springframework.util.ReflectionUtils.findMethod(ReflectionUtils.java:218) ~[spring-core-6.2.11.jar:6.2.11]
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.findBeanMethod(OnBeanCondition.java:688) ~[spring-boot-autoconfigure-3.5.6.jar:3.5.6]
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.getMethodReturnType(OnBeanCondition.java:683) ~[spring-boot-autoconfigure-3.5.6.jar:3.5.6]
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.getReturnType(OnBeanCondition.java:667) ~[spring-boot-autoconfigure-3.5.6.jar:3.5.6]
	at org.springframework.boot.autoconfigure.condition.OnBeanCondition$Spec.deducedBeanTypeForBeanMethod(OnBeanCondition.java:656) ~[spring-boot-autoconfigure-3.5.6.jar:3.5.6]
	... 78 more
Caused by: java.lang.NoClassDefFoundError: io/micrometer/core/instrument/MeterRegistry
```

chillleader committed:
refactor: deprecate old job worker executor methods

Older job worker executor config method is now
replaced with two separate executors.
Marking it as deprecated makes it clear to the users

chillleader committed:
test: fix potential resource leak

We must close the executors in each test case,
if we pass false, they are never closed.

chillleader committed:
test: add new test case to check different executor ownership

chillleader committed:
refactor: set up executors separately

This ensures more clarity in setting up the executors.
The previous code assumed the config impl provides
a default value for the scheduler

chillleader committed:
refactor: rename ExecutorResource to JobWorkerExecutors

This reflects the new nature of this class better
(multiple executors)

chillleader committed:
fix: restore createDefault overloads

they were previously deleted in the same PR but this is not the best idea (breaking change,
need at least an overload without metricsregistry)

chillleader committed:
docs: add javadocs to constructors in CamundaClientExecutorService

chillleader committed:
fix: restore separate spring configuration for executor service

This reverts a change from the same PR that caused
an issue for setups where micrometer dependency
is not present. Now it's again protected by a
ConditionalOnClass

chillleader committed:
fix: restore default configuration for executor service

this was deleted in previous commits but with a
return to the previous setup, the default config is
needed again for cases when there's no micrometer

chillleader committed:
fix: remove redundant metric binding in the configuration

This is now done inside CamundaClientExecutorService

chillleader committed:
fix: extract metered ExecutorService into a separate class

Having the MeterRegistry inside the CamundaClientExecutorService
leads to a NoClassDefFound in setups with no micrometer

chillleader committed:
fix: remove unused logger in CamundaClientExecutorService

chillleader committed:
fix: make MeteredCamundaClientExecutorService null-safe

If micrometer is on the classpath but not configured
in the spring context, MeterRegistry can be null

megglos reviewed (approved):
LGTM! Thanks again for this contribution and addressing all comments!

jonathanlukas reviewed (approved):
Very cool! Thank you 👍

chillleader committed:
fix: remove unused import
